File: typescript/linked-list/linked-list.ts
class LinkedListItem<TElement> {
  public next: LinkedListItem<TElement> | undefined;
  public previous: LinkedListItem<TElement> | undefined;

  constructor(public readonly value: TElement) {}

  public makeOrphan() {
    this.next = undefined;
    this.previous = undefined;
  }

  public unlinkNext() {
    if (!this.next) {
      return;
    }

    this.next.previous = undefined;
    this.next = undefined;
  }

  public linkNext(next: LinkedListItem<TElement> | undefined) {
    if (this.next) {
      throw new Error("Cannot link a new next item, already contains next.");
    }

    this.next = next;

    if (next) {
      next.previous = this;
    }
  }

  public unlinkPrevious() {
    if (!this.previous) {
      return;
    }

    this.previous.next = undefined;
    this.previous = undefined;
  }

  public linkPrevious(previous: LinkedListItem<TElement> | undefined) {
    if (this.previous) {
      throw new Error("Cannot link a new next item, already contains next.");
    }

    this.previous = previous;

    if (previous) {
      previous.next = this;
    }
  }
}

export class LinkedList<TElement> {
  private head: LinkedListItem<TElement> | undefined;
  private tail: LinkedListItem<TElement> | undefined;

  public push(element: TElement) {
    if (this.tail) {
      const item = new LinkedListItem(element);
      this.tail.linkNext(item);
      this.tail = item;
    } else {
      this.head = new LinkedListItem(element);
      this.tail = this.head;
    }
  }

  public pop(): TElement {
    if (this.tail) {
      const result = this.tail.value;

      if (this.tail === this.head) {
        this.tail = undefined;
        this.head = undefined;
      } else {
        const previous = this.tail.previous;
        this.tail.unlinkPrevious();
        this.tail = previous;
      }

      return result;
    } else {
      throw new Error("The list is empty!");
    }
  }

  public shift(): TElement {
    if (this.head) {
      const result = this.head.value;

      if (this.tail === this.head) {
        this.tail = undefined;
        this.head = undefined;
      } else {
        const next = this.head.next;
        this.head.unlinkNext();
        this.head = next;
      }

      return result;
    } else {
      throw new Error("The list is empty!");
    }
  }

  public unshift(element: TElement) {
    if (this.head) {
      const item = new LinkedListItem(element);
      this.head.linkPrevious(item);
      this.head = item;
    } else {
      this.head = new LinkedListItem(element);
      this.tail = this.head;
    }
  }

  public delete(element: TElement) {
    var item = this.head;

    while (item) {
      if (item.value === element) {
        const next = item.next;
        const previous = item.previous;

        item.unlinkPrevious();
        item.unlinkNext();

        next?.linkPrevious(previous);

        if (item === this.head) {
          this.head = next;
        }

        if (item === this.tail) {
          this.tail = previous;
        }
      }

      item = item.next;
    }
  }

  public count(): number {
    let count = 0;
    var item = this.head;

    while (item) {
      count++;
      item = item.next;
    }

    return count;
  }
}

File: go/linked-list/linked_list.go
package linkedlist

import (
	"fmt"
)

type Node struct {
	Value    interface{}
	nextNode *Node
	prevNode *Node
}

type List struct {
	firstNode *Node
	lastNode  *Node
}

func NewList(elements ...interface{}) *List {
	length := len(elements)
	result := List{}

	for i := 0; i < length; i++ {
		result.Push(elements[i])
	}

	return &result
}

func (n *Node) Detach() {
	n.nextNode = nil
	n.prevNode = nil
}

func (n *Node) ReverseBackwards() {
	if n == nil {
		return
	}

	prevNode := n.prevNode
	nextNode := n.nextNode

	prevNode.ReverseBackwards()

	n.nextNode = prevNode
	n.prevNode = nextNode
}

func (n *Node) Next() *Node {
	return n.nextNode
}

func (n *Node) Prev() *Node {
	return n.prevNode
}

func (l *List) Unshift(v interface{}) {
	if l.firstNode == nil {
		l.firstNode = &Node{
			Value: v,
		}
		l.lastNode = l.firstNode
	} else {
		newNode := &Node{
			Value:    v,
			nextNode: l.firstNode,
		}

		l.firstNode.prevNode = newNode
		l.firstNode = newNode
	}
}

func (l *List) Push(v interface{}) {
	if l.lastNode == nil {
		l.firstNode = &Node{
			Value: v,
		}
		l.lastNode = l.firstNode
	} else {
		newNode := &Node{
			Value:    v,
			prevNode: l.lastNode,
		}

		l.lastNode.nextNode = newNode
		l.lastNode = newNode
	}
}

func (l *List) Shift() (interface{}, error) {
	if l.firstNode == nil {
		return nil, fmt.Errorf("cannot shift an empty list")
	}

	firstNode := l.firstNode
	nextNode := firstNode.nextNode

	if nextNode == nil {
		l.firstNode = nil
		l.lastNode = nil
	} else {
		nextNode.prevNode = nil
		l.firstNode = nextNode
	}

	firstNode.Detach()

	return firstNode.Value, nil
}

func (l *List) Pop() (interface{}, error) {
	if l.lastNode == nil {
		return nil, fmt.Errorf("cannot pop an empty list")
	}

	lastNode := l.lastNode
	prevNode := lastNode.prevNode

	if prevNode == nil {
		l.firstNode = nil
		l.lastNode = nil
	} else {
		prevNode.nextNode = nil
		l.lastNode = prevNode
	}

	lastNode.Detach()

	return lastNode.Value, nil
}

func (l *List) Reverse() {
	firstNode := l.firstNode
	lastNode := l.lastNode

	lastNode.ReverseBackwards()

	l.firstNode = lastNode
	l.lastNode = firstNode
}

func (l *List) First() *Node {
	return l.firstNode
}

func (l *List) Last() *Node {
	return l.lastNode
}
